File: airflow/www/static/js/dag.js
/*!
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* global document, window, CustomEvent, $ */

import { getMetaValue } from "./utils";
import { approxTimeFromNow, formatDateTime } from "./datetime_utils";
import { openDatasetModal, getDatasetTooltipInfo } from "./datasetUtils";

const dagId = getMetaValue("dag_id");
const pausedUrl = getMetaValue("paused_url");
// eslint-disable-next-line import/prefer-default-export
export const dagTZ = getMetaValue("dag_timezone");
const datasetsUrl = getMetaValue("datasets_url");
const nextRun = {
  createAfter: getMetaValue("next_dagrun_create_after"),
  intervalStart: getMetaValue("next_dagrun_data_interval_start"),
  intervalEnd: getMetaValue("next_dagrun_data_interval_end"),
};
let nextDatasets = [];
let nextDatasetsError;

const setNextDatasets = (datasets, error) => {
  nextDatasets = datasets;
  nextDatasetsError = error;
};

$(window).on("load", function onLoad() {
  $(`a[href*="${this.location.pathname}"]`).parent().addClass("active");
  $(".never_active").removeClass("active");
  const run = $("#next-dataset-tooltip");
  const singleDatasetUri = $(run).data("uri");
  if (!singleDatasetUri) {
    getDatasetTooltipInfo(dagId, run, setNextDatasets);
  }
});

$("#pause_resume").on("change", function onChange() {
  const $input = $(this);
  const id = $input.data("dag-id");
  const isPaused = $input.is(":checked");
  const requireConfirmation = $input.is("[data-require-confirmation]");
  if (requireConfirmation) {
    const confirmation = window.confirm(
      `Are you sure you want to ${isPaused ? "resume" : "pause"} this DAG?`
    );
    if (!confirmation) {
      $input.prop("checked", !isPaused);
      return;
    }
  }
  const url = `${pausedUrl}?is_paused=${isPaused}&dag_id=${encodeURIComponent(
    id
  )}`;
  // Remove focus on element so the tooltip will go away
  $input.trigger("blur");
  $input.removeClass("switch-input--error");

  // dispatch an event that React can listen for
  const event = new CustomEvent("paused", { detail: isPaused });
  document.dispatchEvent(event);

  $.post(url).fail(() => {
    setTimeout(() => {
      $input.prop("checked", !isPaused);
      $input.addClass("switch-input--error");
      event.value = !isPaused;
      document.dispatchEvent(event);
    }, 500);
  });
});

$("#next-run").on("mouseover", () => {
  $("#next-run").attr("data-original-title", () => {
    let newTitle = "";
    if (nextRun.createAfter) {
      newTitle += `<strong>Run After:</strong> ${formatDateTime(
        nextRun.createAfter
      )}<br>`;
      newTitle += `Next Run: ${approxTimeFromNow(nextRun.createAfter)}<br><br>`;
    }
    if (nextRun.intervalStart && nextRun.intervalEnd) {
      newTitle += "<strong>Data Interval</strong><br>";
      newTitle += `Start: ${formatDateTime(nextRun.intervalStart)}<br>`;
      newTitle += `End: ${formatDateTime(nextRun.intervalEnd)}`;
    }
    return newTitle;
  });
});

$(".next-dataset-triggered").on("click", (e) => {
  const run = $("#next-dataset-tooltip");
  const summary = $(e.target).data("summary");
  const singleDatasetUri = $(run).data("uri");
  if (!singleDatasetUri) {
    openDatasetModal(dagId, summary, nextDatasets, nextDatasetsError);
  } else {
    window.location.href = `${datasetsUrl}?uri=${encodeURIComponent(
      singleDatasetUri
    )}`;
  }
});

File: devel-common/src/tests_common/test_utils/dag.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import annotations

from collections.abc import Collection, Sequence
from typing import TYPE_CHECKING

from airflow.utils.session import NEW_SESSION, provide_session

from tests_common.test_utils.compat import DagSerialization, SerializedDAG

if TYPE_CHECKING:
    from sqlalchemy.orm import Session

    from airflow.sdk import DAG


def create_scheduler_dag(dag: DAG | SerializedDAG) -> SerializedDAG:
    if isinstance(dag, SerializedDAG):
        return dag
    return DagSerialization.deserialize_dag(DagSerialization.serialize_dag(dag))


@provide_session
def sync_dag_to_db(
    dag: DAG,
    bundle_name: str = "testing",
    session: Session = NEW_SESSION,
) -> SerializedDAG:
    return sync_dags_to_db([dag], bundle_name=bundle_name, session=session)[0]


@provide_session
def sync_dags_to_db(
    dags: Collection[DAG],
    bundle_name: str = "testing",
    session: Session = NEW_SESSION,
) -> Sequence[SerializedDAG]:
    """
    Sync dags into the database.

    This serializes dags and saves the results to the database. The serialized
    (scheduler-oeirnted) dags are returned. If the input is ordered (e.g. a list),
    the returned sequence is guaranteed to be in the same order.
    """
    from airflow.models.dagbundle import DagBundleModel
    from airflow.models.serialized_dag import SerializedDagModel
    from airflow.serialization.serialized_objects import LazyDeserializedDAG

    session.merge(DagBundleModel(name=bundle_name))
    session.flush()

    def _write_dag(dag: DAG) -> SerializedDAG:
        data = DagSerialization.to_dict(dag)
        SerializedDagModel.write_dag(LazyDeserializedDAG(data=data), bundle_name, session=session)
        return DagSerialization.from_dict(data)

    SerializedDAG.bulk_write_to_db(bundle_name, None, dags, session=session)
    scheduler_dags = [_write_dag(dag) for dag in dags]
    session.flush()
    return scheduler_dags
